File: Medium/1443. Minimum Time to Collect All Apples in a Tree/MinimumTimeToCollectAllApplesInATree.java
/*
LeetCode Medium: 1443. Minimum Time to Collect All Apples in a Tree
https://leetcode.com/problems/minimum-time-to-collect-all-apples-in-a-tree/

Check MinimumTimeToCollectAllApplesInATree.cpp for brief solution explanation.
*/
class Solution {
    static ArrayList<Integer>[]adj;
    static boolean[]subtreeHasApples;
    static boolean[]vis;
    static int[]sum;

    public int minTime(int N, int[][] edges, List<Boolean> hasApple) {
        adj = new ArrayList[N];
        subtreeHasApples = new boolean[N];
        vis = new boolean[N];
        sum = new int[N];
        for(int i = 0;i<N;i++){
            adj[i] = new ArrayList<Integer>();
        }
        for(int i = 0;i<edges.length;i++){
            adj[edges[i][0]].add(edges[i][1]);
            adj[edges[i][1]].add(edges[i][0]);
        }
        dfs(0, hasApple);
        return sum[0];
    }
    public boolean dfs(int u, List<Boolean> hasApple){
        vis[u] = true;
        if(hasApple.get(u)==true){
            subtreeHasApples[u] = true;
        }
        for(int nxt:adj[u]){
            if(!vis[nxt]){
                boolean t = dfs(nxt, hasApple);
                if(t){
                    sum[u]+=2+sum[nxt];
                }
                subtreeHasApples[u] = subtreeHasApples[u]||t;
            }
        }
        return subtreeHasApples[u];
    }
}

File: Medium/1443. Minimum Time to Collect All Apples in a Tree/MinimumTimeToCollectAllApplesInATree.py
'''
LeetCode Medium: 1443. Minimum Time to Collect All Apples in a Tree
https://leetcode.com/problems/minimum-time-to-collect-all-apples-in-a-tree/

Check MinimumTimeToCollectAllApplesInATree.cpp for brief solution explanation.
'''
class Solution:
    def minTime(self, N: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        adj = [[] for i in range(N)]
        subtreeHasApples = [False for i in range(N)]
        vis = [False for i in range(N)]
        sum = [0 for i in range(N)]
        for i in range(len(edges)):
            adj[edges[i][0]].append(edges[i][1])
            adj[edges[i][1]].append(edges[i][0])
        def dfs(u):
            vis[u] = True
            if hasApple[u]==True:
                subtreeHasApples[u] = True
            for nxt in adj[u]:
                if not vis[nxt]:
                    t = dfs(nxt)
                    if t:
                        sum[u]+=2+sum[nxt]
                    subtreeHasApples[u] = subtreeHasApples[u] or t
            return subtreeHasApples[u]
        dfs(0)
        return sum[0]